orb-ui/cone/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+pub mod button;
+pub mod lcd;
+pub mod led;
+
+use crate::button::Button;
+use crate::lcd::{Lcd, LcdCommand};
+use crate::led::{Led, CONE_LED_COUNT};
+use color_eyre::eyre;
+use color_eyre::eyre::Context;
+use embedded_graphics::pixelcolor::{Rgb565, RgbColor};
+use ftdi_embedded_hal::libftd2xx::{Ft4232h, Ftdi, FtdiCommon};
+use image::{ImageFormat, Luma};
+use orb_rgb::Argb;
+use std::sync::{Arc, Mutex};
+use std::{env, fs};
+use tokio::sync::mpsc;
+
+const CONE_FTDI_DEVICE_COUNT: usize = 8;
+
+#[derive(Debug)]
+enum Status {
+    Connected,
+    Disconnected,
+}
+
+/// Cone can be created only if connected to the host over USB.
+pub struct Cone {
+    connection_status: Arc<Mutex<Status>>,
+    lcd: Lcd,
+    led_strip: Arc<Mutex<Led>>,
+    _button: Button,
+}
+
+pub enum ConeEvents {
+    ButtonPressed(bool),
+}
+
+impl Cone {
+    /// Create a new Cone instance.
+    pub fn new(event_queue: mpsc::UnboundedSender<ConeEvents>) -> eyre::Result<Self> {
+        let connection_status = if ftdi_embedded_hal::libftd2xx::list_devices()?.len()
+            != CONE_FTDI_DEVICE_COUNT
+        {
+            return Err(eyre::eyre!(
+                "FTDI device count mismatch: cone not connected?"
+            ));
+        } else {
+            let mut device: Ft4232h = Ftdi::with_index(6)?
+                .try_into()
+                .wrap_err("Failed to initialize FTDI device")?;
+            device.reset().wrap_err("Failed to reset")?;
+            Arc::new(Mutex::new(Status::Connected))
+        };
+
+        let lcd = Lcd::spawn()?;
+        let led_strip = Led::spawn()?;
+        let _button = Button::spawn(event_queue.clone(), connection_status.clone())?;
+
+        let cone = Cone {
+            connection_status,
+            lcd,
+            led_strip,
+            _button,
+        };
+
+        Ok(cone)
+    }
+
+    pub fn is_connected(&self) -> bool {
```

Comment:
I'm not sure that sharing an arc mutex and storing it in the cone struct like this is a good idea. Probably makes more sense to have Connected and disconnected events that are sent through the channel.

Alternatively, use an arc [atomic enum](https://docs.rs/atomic_enum/latest/atomic_enum/) or an `Arc<AtomicBool>`. You won't have to think about deadlocks or poisoning that way.

Reply:
done